popup.js

```diff
@@ -135,45 +141,53 @@ function minimizeStyles(message) {
  * @param {Document} doc The Document that contains the rendered HTML.
  * @param {Element} element The Element whose style attributes should be
  *     minimized.
+ * @param {string} id The id of the Element in the final page.
  * @param {number} index The index in |message.html| where the Element's style
  *     attribute is specified.
  */
-function minimizeStyle(message, doc, element, index) {
-  var originalStyle = element.getAttribute('style');
-  element.removeAttribute('style');
-  var unstyledStyle = doc.defaultView.getComputedStyle(element, null);
-
-  var fullStyleDeclaration = message.html[index];
-  var stylePrefix;
-  var styleSuffix;
-  // Remove style declaration from style attribute.
-  var style = fullStyleDeclaration.replace(
-    /^style=("|&(amp;)*?quot;)(.*)("|&(amp;)*?quot;) ?$/,
-    function(match, quote, p2, content) {
-      stylePrefix = 'style=' + quote;
-      styleSuffix = quote + ' ';
-      return content;
+function minimizeStyle(message, doc, element, id, index) {
+  var originalStyleAttribute = element.getAttribute('style');
+  var originalStyleMap = message.idToStyleMap[id];
+  var requiredStyleMap = {};
+  var maxNumberOfIterations = 5;
+
+  for (var i = 0; i < maxNumberOfIterations; i++) {
+    var currentStyleAttribute = [];
+    for (var property in requiredStyleMap) {
+      currentStyleAttribute.push(
+        property + ':' + requiredStyleMap[property] + ';'
+      );
     }
-  );
-  // Remove property value pair if unstyledStyle has the same property
-  // value.
-  style = style.replace(
-    /([\S]*?): ("|&(amp;)*?quot;)?(.*?)("|&(amp;)*?quot;)?; ?/g,
-    function(match, property, p2, p3, value) {
-      unstyledValue = unstyledStyle[property].replace(/"/g, '');
-      if (unstyledValue == value) {
-        return '';
-      } else {
-        return match;
+    element.setAttribute('style', currentStyleAttribute.join(' '));
+    var currentComputedStyle = doc.defaultView.getComputedStyle(element, null);
+    var foundNewRequiredStyle = false;
+    for (var property in originalStyleMap) {
+      var originalValue = originalStyleMap[property];
+      if (originalValue != currentComputedStyle.getPropertyValue(property)) {
+        requiredStyleMap[property] = originalValue;
+        foundNewRequiredStyle = true;
       }
     }
-  );
+    if (!foundNewRequiredStyle) {
+      break;
+    }
+  }
+  element.setAttribute('style', originalStyleAttribute);
+
+  var finalStyleAttribute = [];
+  for (var property in requiredStyleMap) {
+    finalStyleAttribute.push(property + ':' + requiredStyleMap[property] + ';');
+  }
+  var style = finalStyleAttribute.join(' ');
+
   if (style) {
-    message.html[index] = stylePrefix + style + styleSuffix;
+    var nestingDepth = message.frameIndex.split('.').length - 1;
+    style = style.replace(/"/g, escapedQuote(nestingDepth + 1));
```

Comment:
I only escape quotes because (as far as I can tell) there won't be any other special characters in the style declaration.  There will only be letters, numbers, dashes, and quotes.  Some values can have quotes which is why I escape them.  Is this not the case?